File: Java/GRPCStreamingTest/server/src/main/java/app/services/ComputeSupportService.java
package app.services;

import java.util.Random;

import computation.Computation.DataItem;
import computation.Computation.DataRange;
import computation.ComputeSupportGrpc.ComputeSupportImplBase;
import io.grpc.stub.StreamObserver;

public class ComputeSupportService extends ComputeSupportImplBase {
    
    private static final Random rdm = new Random();

    @Override
    public void generateValues(DataRange request, StreamObserver<DataItem> responseObserver) {
		double min = request.getLower();
		double max = request.getUpper();
        for(int i = 0; i < request.getCount(); ++i){
            double sample = rdm.nextDouble(min, max);
            responseObserver.onNext(DataItem.newBuilder().setValue(sample).build());
        }
        responseObserver.onCompleted();
    }

    @Override
    public StreamObserver<DataItem> combineValues(StreamObserver<DataItem> responseObserver) {
        return new StreamObserver<DataItem>() {
            
            double sum = 0;

            @Override
            public void onCompleted() {
                double result = Math.sqrt(sum);
                responseObserver.onNext(DataItem.newBuilder().setValue(result).build());
                responseObserver.onCompleted();
            }

            @Override
            public void onError(Throwable t) {
                t.printStackTrace();
            }

            @Override
            public void onNext(DataItem entry) {
                double value = entry.getValue();
                sum += value * value;
            }
        };
    }
    
}

File: DotNet/GRPCStreamingTest/ServerApp/Services/ComputeSupportService.cs
using Computation;
using Grpc.Core;

namespace ServerApp.Services;

public class ComputeSupportService : ComputeSupport.ComputeSupportBase
{
    public override async Task GenerateValues(DataRange request, IServerStreamWriter<DataItem> responseStream, ServerCallContext context)
    {
        double min = request.Lower;
        double max = request.Upper;
        for(int i = 0; i < request.Count; ++i)
        {
            double sample = min + (max - min) * Random.Shared.NextDouble();
            await responseStream.WriteAsync(new DataItem { Value = sample });
        }
    }

    public override async Task<DataItem> CombineValues(IAsyncStreamReader<DataItem> requestStream, ServerCallContext context)
    {
        double sum = 0;
        while(await requestStream.MoveNext())
        {
            double value = requestStream.Current.Value;
            sum += value * value;
        }
        return new DataItem { Value = Math.Sqrt(sum) };
    }
}
